src/Analyser/TypeSpecifier.php

```diff
@@ -1105,10 +1102,8 @@ private function specifyTypesForCountFuncCall(
 			if (
 				$sizeType instanceof IntegerRangeType
 				&& $sizeType->getMin() !== null
-				&& (
-					$isList->yes()
-					|| $isConstantArray->yes() && $arrayType->getKeyType()->isSuperTypeOf(IntegerRangeType::fromInterval(0, $sizeType->getMin() - 1))->yes()
-				)
+				&& ($sizeType->getMax() ?? $sizeType->getMin()) < ConstantArrayTypeBuilder::ARRAY_COUNT_LIMIT
```

Comment:
I don't think that `($sizeType->getMax() ?? $sizeType->getMin())` is the final size of the array created here.

Please refactor this section. You can prepare the arguments for `setOffsetValueType` in a single array beforehand, count the number of elements, and then proceed with creating the constant array only if it's below `ConstantArrayTypeBuilder::ARRAY_COUNT_LIMIT`.

Reply:
I still think it does, I tried to describe the logic here https://github.com/phpstan/phpstan-src/pull/3902#discussion_r2012456960

but you can also check https://github.com/phpstan/phpstan-src/blob/dfd423fdea473b5d0c50bf21f3f560a777e9a6d9/src/Analyser/TypeSpecifier.php#L1108-L1133

case (for loop) 1 and 2 combined call `setOffsetValueType()` up to `max` times. case 3 is special but limited by an existing constant array which should make it safe.

do you disagree? I don't even know how to refactor this right now tbh, I think the "max if set, otherwise min" is the simplest way of dealing with this.